Add section about image cleaning for disk space (#348)
harvester/docs: docs/faq.md

### I added an additional disk with partitions. Why is it not getting detected?

As of Harvester v1.0.2, we no longer support adding additional partitioned disks, so be sure to delete all partitions first (e.g., using `fdisk`).

### Why are there some Harvester pods that become ErrImagePull/ImagePullBackOff?

> This is likely because your Harvester cluster is an air-gapped setup, and some pre-loaded container images are missing. Kubernetes has a mechanism that does garbage collection against bloated image stores. When the partition which stores container images is over 85% full, `kubelet` will try to prune some least used images to save disk space until the occupancy is lower than 80%. These numbers (85% and 80%) are default High/Low thresholds that come with Kubernetes.

Reviewer: ```suggestion
If Harvester pods show `ErrImagePull` or `ImagePullBackOff`, your Harvester cluster is likely an air-gapped setup, and some pre-loaded container images are missing. Kubernetes has a mechanism that does garbage collection against bloated image stores. When the partition storing container images is over 85% full, `kubelet` tries to prune images that are no longer being used to save disk space until the occupancy of the image store is lower than 80%. The 85% and 80% numbers are the default High/Low thresholds in Kubernetes.
```

Author: Per the sentence

> ... tries to prune images that are no longer being used ...

And the paragraph from the [Kubernetes documentation page](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#container-image-lifecycle)

> Disk usage above the configured `HighThresholdPercent` value triggers garbage collection, which deletes images in order based on the last time they were used, starting with the oldest first.

An image may still be being used but pruned by kubelet. I suggest going with the official docs, i.e., changing the sentence from "no longer being used ..." to "the images based on the last time they were used, starting with the oldest," to clarify.

```suggestion
This is likely because your Harvester cluster is an air-gapped setup, and some pre-loaded container images are missing. Kubernetes has a mechanism that does garbage collection against bloated image stores. When the partition which stores container images is over 85% full, `kubelet` tries to prune the images based on the last time they were used, starting with the oldest, until the occupancy is lower than 80%. These numbers (85% and 80%) are default High/Low thresholds that come with Kubernetes.
```

What do you think?

Reviewer: @starbops Agreed, want to the official K8s docs. Thanks for calling this out.

> To recover from the state, either pull the missing images from the Internet (by setting up an HTTP proxy if it's an air-gapped environment) or manually import the images from the ISO image. For multi-node clusters, chances are the missing container images on one node could be found on the other nodes. If that's the case, you can export the images from the node that still has them and import them on the missing one.

Reviewer: ```suggestion
To recover from this state for a single node cluster, do one of the following:  
    - If it's an air-gapped environment, pull the missing images from the Internet by setting up an HTTP proxy.
    - Manually import the images from the ISO image. 
    
For multi-node clusters, you can likely find the missing container images on one node on the other nodes. If that's the case, you can export the images from the node still with them and import them on the missing one.
```

Author: The two bullet points are not applied to single-node cases only. And "finding missing container images on the other nodes" implies it's a multi-node case. So I'd like to combine your suggestions and rewrite it like the following:
```suggestion
To recover from this state, do one of the following depending on the cluster's configuration:
- Pull the missing images from sources outside of the cluster (if it's an air-gapped environment, you might need to set up an HTTP proxy beforehand)
- Manually import the images from the Harvester ISO image
- Find the missing images on one node on the other nodes, and then export the images from the node still with them and import them on the missing one
```
Thank you!

> To prevent such things from happening, please try to clean up unused container images from the previous version each time after a successful Harvester upgrade if the image store disk space is stressed. We provided a handy script that helps clean up disk space, especially for container storage, which can be found [here](https://github.com/harvester/upgrade-helpers/blob/main/bin/harv-purge-images.sh). The script has to be executed on each Harvester node. For example, if the cluster was originally in v1.1.2, and now it gets upgraded to v1.2.0, you can do the following to discard the container images that are only used in v1.1.2 but no longer needed in v1.2.0:

Reviewer: ```suggestion
To prevent this from happening, we recommend cleaning up unused container images from the previous version after each successful Harvester upgrade if the image store disk space is stressed. We provided a [harv-purge-image script](https://github.com/harvester/upgrade-helpers/blob/main/bin/harv-purge-images.sh) that makes cleaning up disk space easy, especially for container storage. The script has to be executed on each Harvester node. For example, if the cluster was originally in v1.1.2, and now it gets upgraded to v1.2.0, you can do the following to discard the container images that are only used in v1.1.2 but no longer needed in v1.2.0:
```

```shell
# on each node
$ ./harv-purge-images.sh v1.1.2 v1.2.0
```

:::caution

> - The script only downloads the image lists and compares the two to calculate the diff between the two versions. It does not check with the cluster and does not know what version the cluster was upgraded from.

Reviewer: ```suggestion
- The script only downloads the image lists and compares the two to calculate the difference between the two versions. It does not communicate with the cluster and, as a result, does not know what version the cluster was upgraded from.
- We published image lists for each version released since v1.1.0. For clusters older than v1.1.0, users must manually clean up the old images.
```

### I added an additional disk with partitions. Why is it not getting detected?

As of Harvester v1.0.2, we no longer support adding additional partitioned disks, so be sure to delete all partitions first (e.g., using `fdisk`).

### Why are there some Harvester pods that become ErrImagePull/ImagePullBackOff?

This is likely because your Harvester cluster is an air-gapped setup, and some pre-loaded container images are missing. Kubernetes has a mechanism that does garbage collection against bloated image stores. When the partition which stores container images is over 85% full, `kubelet` tries to prune the images based on the last time they were used, starting with the oldest, until the occupancy is lower than 80%. These numbers (85% and 80%) are default High/Low thresholds that come with Kubernetes.

To recover from this state, do one of the following depending on the cluster's configuration:
- Pull the missing images from sources outside of the cluster (if it's an air-gapped environment, you might need to set up an HTTP proxy beforehand).

> - Manually import the images from the Harvester ISO image.

Reviewer 2: can u specify how to do that or link to a proper doc/script?

Author: I've updated it with a side note about how to achieve that. PTAL, thanks.

> - Pull the missing images from sources outside of the cluster (if it's an air-gapped environment, you might need to set up an HTTP proxy beforehand)
> - Manually import the images from the Harvester ISO image
> - Find the missing images on one node on the other nodes, and then export the images from the node still with them and import them on the missing one

Reviewer: ```suggestion
- Pull the missing images from sources outside of the cluster (if it's an air-gapped environment, you might need to set up an HTTP proxy beforehand).
- Manually import the images from the Harvester ISO image.
- Find the missing images on one node on the other nodes, and then export the images from the node still with them and import them on the missing one.
```

> - Pull the missing images from sources outside of the cluster (if it's an air-gapped environment, you might need to set up an HTTP proxy beforehand)
> - Manually import the images from the Harvester ISO image
> - Find the missing images on one node on the other nodes, and then export the images from the node still with them and import them on the missing one

Reviewer: ```suggestion
- Pull the missing images from sources outside of the cluster (if it's an air-gapped environment, you might need to set up an HTTP proxy beforehand).
- Manually import the images from the Harvester ISO image.
- Find the missing images on one node on the other nodes, and then export the images from the node still with them and import them on the missing one.
```

> - Find the missing images on one node on the other nodes, and then export the images from the node still with them and import them on the missing one.

Reviewer 2: ```suggestion
- Find the missing images on that node from the other nodes, then export the images from the node where the images still exist and import it on the missing node.
```

To prevent this from happening, we recommend cleaning up unused container images from the previous version after each successful Harvester upgrade if the image store disk space is stressed. We provided a [harv-purge-images script](https://github.com/harvester/upgrade-helpers/blob/main/bin/harv-purge-images.sh) that makes cleaning up disk space easy, especially for container image storage. The script has to be executed on each Harvester node. For example, if the cluster was originally in v1.1.2, and now it gets upgraded to v1.2.0, you can do the following to discard the container images that are only used in v1.1.2 but no longer needed in v1.2.0:

```shell
# on each node
$ ./harv-purge-images.sh v1.1.2 v1.2.0
```

:::caution

- The script only downloads the image lists and compares the two to calculate the difference between the two versions. It does not communicate with the cluster and, as a result, does not know what version the cluster was upgraded from.

> - We published image lists for each version released since v1.1.0. For clusters older than v1.1.0, users have to clean up the old images manually.

Reviewer: ```suggestion
- We published image lists for each version released since v1.1.0. For clusters older than v1.1.0, you have to clean up the old images manually.
```
